### ISSUE TITLE
Custom elements with readonly attributes (like Polymer) fail

### ISSUE THREAD

**Issue by naturalethic** [lhorie/mithril.js#924]
As Polymer creates properties on custom components that Mithril may not be aware of, Mithril should take care to finally setAttribute if node[attr] = val fails.

Perhaps like this (from line 579)
```javascript
					else if (attrName in node && attrName !== "list" && attrName !== "style" && attrName !== "form" && attrName !== "type" && attrName !== "width" && attrName !== "height") {
						//#348 don't set the value if not needed otherwise cursor placement breaks in Chrome
						try {
							if (tag !== "input" || node[attrName] !== dataAttr) node[attrName] = dataAttr;
						} catch (e) {
							node.setAttribute(attrName, dataAttr);
						}
					}
					else node.setAttribute(attrName, dataAttr);
```

**Comment by lhorie** [comment]
merged PR. Thanks!

**Comment by barneycarroll** [comment]
TL;DR: this is the job of the custom element API, not Mithril's.

This illustrates an interesting problem for isomorphism. HTML > DOM relies on the fact that there are no shared keys between attributes and properties — in other words, having distinct attributes and properties with the same key "breaks the web". If Mithril sees that a given key is available as a DOM property, and that property can't be set, I would surmise that that the author has misunderstood the custom element's API.

Native DOM element properties are usually getter / setters under the hood. Take `href` or `type` for example: what appears to be direct assignment and reading of a string in HTML is in fact a complex process. If the assignment (setter function) is useless, should the underlying API throw, or silently noop? That's an issue for the element in question. It's impossible for Mithril to have a 'right' opinion on how to second guess this.

If I play devil's advocate for a second, we might look to prior form and see that `delete object[ key ]` is an expression that will evaluate to `false` if it failed to achieve the presumed intent. So we might argue that Mithril should follow that convention and parse the response of the attempted assignment to see if it behaves similarly and fall back to `setAttribute`. Does the underlying element API conform to this expectation? Or does it behave in the same way that an attempt to change an `input` type, by throwing a descriptive error? Mithril isn't qualified to answer these questions — it should assume nothing and let the element's API do the thinking.

**Comment by isiahmeadows** [comment]
@barneycarroll Still, I don't think that we should assume that all custom elements are well-behaved, so for interoperability reasons, we should still use that fallback when necessary. It's not Mithril's fault if a custom component doesn't fulfill the traditional element contract (e.g. `input` only has a getter), but it most certainly looks like it to someone not as familiar with Mithril's internals. I would expect similar reactions from people if they had the same problem with, say, React or Angular 2.

For the same reasons, [source-map has to accept invalid source maps to play nice with Sass 3.3](https://github.com/mozilla/source-map/blob/master/lib/source-map-consumer.js#L290-L302), specifically because they don't emit a valid source map. But to those not as familiar with how source maps are handled, it looks like a problem in the browser or whatever plugin.

**Comment by barneycarroll** [comment]
Angular 2 FTW!
